pkg/controllers/dynakube/injection/reconciler.go

```diff
@@ -0,0 +1,170 @@
+package injection
+
+import (
+	"context"
+	goerrors "errors"
+
+	"github.com/Dynatrace/dynatrace-operator/pkg/api/status"
+	dynatracev1beta1 "github.com/Dynatrace/dynatrace-operator/pkg/api/v1beta1/dynakube"
+	"github.com/Dynatrace/dynatrace-operator/pkg/clients/dynatrace"
+	"github.com/Dynatrace/dynatrace-operator/pkg/controllers"
+	"github.com/Dynatrace/dynatrace-operator/pkg/controllers/dynakube/istio"
+	"github.com/Dynatrace/dynatrace-operator/pkg/controllers/dynakube/version"
+	"github.com/Dynatrace/dynatrace-operator/pkg/injection/namespace/ingestendpoint"
+	"github.com/Dynatrace/dynatrace-operator/pkg/injection/namespace/initgeneration"
+	"github.com/Dynatrace/dynatrace-operator/pkg/injection/namespace/mapper"
+	"github.com/Dynatrace/dynatrace-operator/pkg/util/timeprovider"
+	"github.com/pkg/errors"
+	"github.com/spf13/afero"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type reconciler struct {
+	client            client.Client
+	apiReader         client.Reader
+	dynakube          *dynatracev1beta1.DynaKube
+	istioReconciler   istio.Reconciler
+	versionReconciler version.Reconciler
+}
+
+type ReconcilerBuilder func(
+	client client.Client,
+	dynatraceClient dynatrace.Client,
+	istioClient *istio.Client,
+	apiReader client.Reader,
+	fs afero.Afero,
+	dynakube *dynatracev1beta1.DynaKube,
+) controllers.Reconciler
+
+//nolint:revive
+func NewReconciler(
+	client client.Client,
+	dynatraceClient dynatrace.Client,
+	istioClient *istio.Client,
+	apiReader client.Reader,
+	fs afero.Afero,
+	dynakube *dynatracev1beta1.DynaKube,
+) controllers.Reconciler {
+	var istioReconciler istio.Reconciler = nil
+
+	if istioClient != nil {
+		istioReconciler = istio.NewReconciler(istioClient)
+	}
+
+	return &reconciler{
+		client:            client,
+		apiReader:         apiReader,
+		dynakube:          dynakube,
+		istioReconciler:   istioReconciler,
+		versionReconciler: version.NewReconciler(apiReader, dynatraceClient, fs, timeprovider.New().Freeze()),
+	}
+}
+
+func (r *reconciler) Reconcile(ctx context.Context) error {
+	if r.istioReconciler != nil {
+		err := r.istioReconciler.ReconcileAPIUrl(ctx, r.dynakube)
```

Comment:
```suggestion
		err := r.istioReconciler.ReconcileCodeModuleCommunicationHosts(ctx, r.dynakube)
```

you reconcile the `APIUrl` in the dynakube-controller as its a general property, and anything that will want to use the dtclient will have that as a prerequisite

Reply:
but actually you already do that at: https://github.com/Dynatrace/dynatrace-operator/pull/2736/files#diff-1f094421eabf0cb587da250be2bbd4bba051c5e50cd36d79d48358ff685171afR125

so it actually not necessary

Reply:
:+1: I will remove it then.

I'm not 100% familiar with how it works. In my mind, as the end goal is to make the reconcilers more independent, I wanted to execute in injection reconciler the same functions that are triggered on active gate reconciler.

e.g. In case injection reconciler is executed 'alone', `func ReconcileAPIUrl` won't be executed because it is done in the dynakube controller, but I guess there is no use case where dynakube reconciler isn't executed alongside injection reconciler, so it makes sense to not repeat `func ReconcileAPIUrl`.

Reply:
[Fixed here](https://github.com/Dynatrace/dynatrace-operator/pull/2736/commits/5b89aab62c04646dadaffe0d57f2d85009c1b74c)